File: yukicoder/p6923/bfs.py
import queue

infl = 1000000000

n, m, k = map(int, input().split())
a, b = map(int, input().split())
g = [[] for _ in range(n)]
for i in range(m):
    x, y = map(int, input().split())
    g[x - 1].append(y - 1)
    g[y - 1].append(x - 1)

q = queue.Queue()
q.put(a - 1)
depth = [infl for _ in range(n)]
depth[a - 1] = 0
visited = [False for _ in range(n)]
visited[a - 1] = True

while not q.empty():
    dest = q.get()
    for v in g[dest]:
        if not visited[v]:
            visited[v] = True
            depth[v] = depth[dest] + 1
            q.put(v)

if k == depth[b - 1]:
    print("shortest")
else:
    print(k - depth[b - 1])

File: yukicoder/p6923/bfs.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
  int n, m, k; cin >> n >> m >> k;
  int a, b; cin >> a >> b;
  a--, b--;
  vector<vector<int>> g(n, vector<int>(0));
  for(int i = 0; i < m; i++){
    int x, y; cin >> x >> y;
    x--, y--;
    g[x].push_back(y);
    g[y].push_back(x);
  }

  queue<int> q;
  q.push(a);
  vector<int> depth(n, INT_MAX);
  depth[a] = 0;
  vector<bool> visited(n, false);
  visited[a] = true;

  while(!q.empty()){
    int dest = q.front();
    q.pop();
    for(int v : g[dest]){
      if(!visited[v]){
        visited[v] = true;
        depth[v] = depth[dest] + 1;
        q.push(v);
      }
    }
  }
  if(k == depth[b]){
    cout << "shortest" << endl;
  }else{
    cout << k - depth[b] << endl;
  }
}